config.org
#+TITLE: w8stes's GNU Emacs Config
#+AUTHOR: w8ste
#+STARTUP: showerverything
#+OPTIONS: toc:2
* TABLE OF CONTENTS :toc:
- [[#important-programs][Important programs]]
  - [[#startup-performance][Startup Performance]]
  - [[#package-system][Package System]]
  - [[#general-settings][General settings]]
  - [[#keybinds][Keybinds]]
- [[#calender][Calender]]
- [[#diminish][Diminish]]
- [[#all-the-icons][All the icons]]
- [[#auctex][AUCTex]]
- [[#autopair][Autopair]]
- [[#move-text][Move-Text]]
- [[#dired][Dired]]
- [[#flycheck][Flycheck]]
- [[#eshell][Eshell]]
- [[#fonts][FONTS]]
  - [[#nerd-icons][Nerd Icons]]
  - [[#settings-the-font-face][Settings The Font Face]]
- [[#git-programms][Git Programms]]
  - [[#magit][Magit]]
- [[#gui-settings][GUI Settings]]
  - [[#disable-menubar-toolbars-and-scrollbars][Disable Menubar, Toolbars and Scrollbars]]
  - [[#display-line-numbers-and-truncated-lines][Display Line Numbers and truncated Lines]]
- [[#gradle-mode][Gradle-Mode]]
- [[#highlight-indents][Highlight Indents]]
- [[#highlight][Highlight]]
- [[#identation][Identation]]
- [[#ivy-counsel][IVY (Counsel)]]
- [[#language-support][Language Support]]
  - [[#lsp-mode][Lsp mode]]
  - [[#company][Company]]
  - [[#java][Java]]
  - [[#python][Python]]
  - [[#julia][Julia]]
  - [[#system-verilog][System Verilog]]
  - [[#c][C++]]
  - [[#haskell][Haskell]]
  - [[#ltex][ltex]]
  - [[#scala][Scala]]
  - [[#dart][Dart]]
- [[#minibuffer-escape][Minibuffer Escape]]
  - [[#enable-toc][Enable ToC]]
  - [[#enable-bullet-points][Enable bullet points]]
  - [[#disable-electric-indent][Disable electric indent]]
- [[#multiple-cursor][Multiple Cursor]]
- [[#pdf-tools][PDF Tools]]
- [[#org-mode-packages][Org Mode Packages]]
  - [[#org-noter][Org Noter]]
  - [[#org-pdftools][Org PDFtools]]
- [[#perspective][Perspective]]
- [[#rainbow-delimiters][Rainbow Delimiters]]
- [[#rainbow-mode][Rainbow Mode]]
  - [[#eshell-1][Eshell]]
- [[#theme][Theme]]
- [[#treesitter][Treesitter]]
- [[#which-key][WHICH-KEY]]
- [[#yasnippet][Yasnippet]]

* Important programs
** Startup Performance
#+begin_src emacs-lisp
(setq gc-cons-threshold (* 50 1000 1000))

(defun efs/display-startup-time ()
  (message "Emacs loaded in %s with %d garbage collections."
           (format "%.2f seconds"
                   (float-time
                     (time-subtract after-init-time before-init-time)))
           gcs-done))

(add-hook 'emacs-startup-hook #'efs/display-startup-time)
#+end_src
** Package System
#+begin_src emacs-lisp
(require 'package)

(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))

(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))

  ;; Initialize use-package on non-Linux platforms
(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)
#+end_src


** General settings
#+begin_src emacs-lisp
    ;; Disable backup files.
    (setq make-backup-files nil)
    ;; Prompt to delete autosaves when killing buffers.
    (setq kill-buffer-delete-auto-save-files t)

    (global-visual-line-mode 0)
    (toggle-truncate-lines 0)
    (setq inhibit-splash-screen t)
    (setq warning-minimum-level :error)  ; Only show errors, not warnings
    (setq warning-suppress-types '((comp)))  ; Suppress all compilation warnings
#+end_src

** Keybinds
#+begin_src emacs-lisp
  ;; magit
  (keymap-global-set "C-c g" 'magit)

  ;; eshell
  (keymap-global-set "C-c e" 'eshell)

  ;; buffer
  ;(which-key-add-key-based-replacements "C-c b" "buffer")

  (keymap-global-set "C-c b i" 'ibuffer)
  (keymap-global-set "C-c b p" 'previous-buffer)
  (keymap-global-set "C-c b n" 'next-buffer)

  ;; Text scale
  ;(which-key-add-key-based-replacements "C-c s" "scale")
  (keymap-global-set "C-c s i" 'text-scale-increase)
  (keymap-global-set "C-c s d" 'text-scale-decrease)

  ;; compile
  (keymap-global-set "C-c c" 'compile)

  ;; tooling
  ;(which-key-add-key-based-replacements "C-c t" "tools")
  (keymap-global-set "C-c t s" 'sbt-start)
  (keymap-global-set "C-c t n" 'sbt-switch-to-active-sbt-buffer)

  (keymap-global-set "C-c r" 'rename-file)
  #+end_src

* Calender
#+begin_src emacs-lisp
(setq calendar-week-start-day 1)
#+end_src

* Diminish
Option that implements hiding of minor-modes in the modeline
#+begin_src emacs-lisp
    (use-package diminish)
    (diminish 'projectile-mode)
#+end_src 


* All the icons
#+begin_src emacs-lisp
  (use-package all-the-icons
    :ensure t
    :if (display-graphic-p))

  (use-package all-the-icons-dired
    :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))
  (add-hook 'dired-mode-hook 'dired-omit-mode)
#+end_src

* AUCTex

Latex LSP Server install: cargo install --locked --git https://github.com/latex-lsp/texlab.git
#+begin_src emacs-lisp
(use-package auctex
  :defer t
  :mode ("\\.tex\\'" . tex-mode)
  :hook (tex-mode . lsp-deferred)
  :config
  (setq TeX-source-correlate-start-server t))

;; Make sure this path matches the one installed via cargo
(setenv "PATH" (concat (expand-file-name "~/.cargo/bin") ":" (getenv "PATH")))
(add-to-list 'exec-path (expand-file-name "~/.cargo/bin"))

#+end_src

* Autopair
#+begin_src emacs-lisp
  (use-package smartparens
    :init
    (smartparens-global-mode))
#+end_src

* Move-Text
#+begin_src emacs-lisp
  (use-package move-text
  :ensure t
  :config
  (move-text-default-bindings))  ;; binds M-↑ and M-↓
#+end_src

* Dired
#+begin_src emacs-lisp
  (use-package dired-open
    :config
    (setq dired-open-extensions '(("gif" . "sxiv")
                                  ("jpg" . "sxiv")
                                  ("png" . "sxiv")
                                  ("mkv" . "mpv")
                                  ("mp4" . "mpv"))))
#+end_src

* Flycheck
#+begin_src emacs-lisp

    (use-package flycheck
    :ensure t
    :init (global-flycheck-mode)
    :config
    (setq flycheck-display-errors-function
      #'flycheck-display-error-messages-unless-error-list)

    (setq flycheck-indication-mode nil))

  (use-package flycheck-pos-tip
    :ensure t
    :after flycheck
    :config
    (flycheck-pos-tip-mode))
#+end_src

* Eshell
#+begin_src emacs-lisp
  (defun make-shell (name)
    "Create a shell buffer named NAME."
    (interactive "sName: ")
    (setq name (concat "$" name))
    (eshell)
    (rename-buffer name))

#+end_src

* FONTS
** Nerd Icons
#+begin_src emacs-lisp
  (use-package nerd-icons)
#+end_src
** Settings The Font Face
#+begin_src emacs-lisp
  (set-frame-font "FiraCode Nerd Font 19")
  (set-face-attribute 'default nil
                      :font "FiraCode Nerd Font"
                      :height 110
                      :weight 'medium)
  (set-face-attribute 'variable-pitch nil
                      :font "FiraCode Nerd Font"
                      :height 120
                      :weight 'medium)
  (set-face-attribute 'fixed-pitch nil
                      :font "FiraCode Nerd Font"
                      :height 110
                      :weight 'medium)
  ;; Makes commented text and keywords italics.
  ;; This is working in emacsclient but not emacs.
  ;; Your font must have an italic face available.
  (set-face-attribute 'font-lock-comment-face nil
                      :slant 'italic)
  (set-face-attribute 'font-lock-keyword-face nil
                      :slant 'italic)

  ;; This sets the default font on all graphical frames created after restarting Emacs.
  ;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
  ;; are not right unless I also add this method of setting the default font.

eshell/history
g++ main.cpp -o main
main
git status
git add config.el config.org
git status
git status
git commit -m "Added doom-modeline"
git push origin main
cd ..
git add em
git add emacs
git commit
cd ..
git commit -m "Update emacs"
git push origin main
exa -lh
g++ main.cpp -o main
neofetch
clear
g++ -o main main.cpp 
g++ -o main main.cpp 
g++ -o main main.cpp 
g++ -o main main.cpp 
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main
ls 
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
g++ main.cpp -o main
/etc/redhat-release
/etc/arch-release
g++ main.cpp -o main
g++ main.cpp -o main
g++ --version
g++ -std=c++17 main.cpp
cat /etc/arch-release 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
g++ main.cpp -o main 
./main
/bin/synth.sh registerFile.sv 
/bin/synth.sh registerFile
zathura registerFile.pdf
/bin/synth.sh regFil
mkdir blocks
mv registerFile.* blocks/ 
rm -rf test
git init 
git status
ls 
yosys opt -fine adder 
yosys
rm adder.pdf
rm adder.dot
mv registerFile.* blocks/ 
yosys -h
yosys
/bin/synth.sh adder
/bin/synth.sh registerFile
yosys
/bin/synth.sh adder
cd ..
mkdir .images
/bin/synth.sh multiplexer
/bin/synth.sh registerFile
mv registerFile.pdf ../.images/
cd blocks
rm *.pdf
rm *.*~
rm iberty 
rm show 
/bin/synth.sh resetenff
/bin/synth.sh resetff
rm *.pdf
rm iberty 
/bin/synth.sh resetff
rm *.pdf
rm *.dot
cd ..
cd block
cd blocks/ 
/bin/synth.sh alu.sv 
/bin/synth.sh alu
/bin/synth.sh decoder
cd ..
/bin/synth.sh datapath
cd main/ 
/bin/synth.sh datapath
mv ../block/alu .
cd ..
mv blocks/alu.sv main
cd main/ 
/bin/synth.sh datapath
cd ..
mv blocks/multiplexer.sv main
cd main/ 
/bin/synth.sh datapath
mv -r main/ block/
mv -r main/* block/
mv  main/* block/
mv main block
cd ..
mv main/ block
cd block
cd ,,
No such directory: ,,
mv main blocks  
cd ..
mv main blocks  
/usr/bin/mv: invalid option -- 'r'
mv block blocks  
cd blocks/
cd block/ 
mv --help
mv . ..
mv *.sv ..
cd ..
rm -rf block
/bin/synth.sh datapath
mv datapath.pdf ../.images
cd ..
mv blocks/datapath.sv blocks/decoder.sv main
mv blocks/datapath.sv blocks/decoder.sv /main/
mv blocks/datapath.sv blocks/decoder.sv ./main/
mkdir main
mv blocks/datapath.sv blocks/decoder.sv /main/
mv blocks/datapath.sv blocks/decoder.sv main/
/bin/synth.sh condcheck
/bin/synth.sh condlogic
mv alu.sv ..
/bin/synth.sh condlogic
mv *.sv ../blocks 
cd ../blocks
/bin/synth.sh condlogic
/bin/synth.sh condcheck
cd ..
mv alu block
cd alu.sv block
mv alu.sv blocks
cd blocks
/bin/synth.sh alu
mv *.pdf ../.images/
mv alu.sv condcheck.sv condlogic.sv datapath.sv decoder.sv controller.sv ../main/
mv *.sv ../blocks
cd ../blocks/ 
.rw-r--r--   595 hannes 16 Oct 09:44  alu.sv
/bin/synth.sh datamem
/bin/synth.sh alu
/bin/synth.sh instrmem
rm -rf .
rm -rf *
ls 
rm exercise01.tex 
mkdir 01
mv /
cd pictures/ 
cd ,,
cd ..
rmdir pictures/ 
mv myTemplate.tex exercise1.tex
mv exercise1.tex 01
rm -rf *
mkdir 01 
mv pictures/ 01
mv myTemplate.tex 01/exercise1.tex
cd 01
cd ..
lh
cd ..
cd ,,
cd ..
mv latexmkrc DT/Tutor/01/
rm projectile-bookmarks.eld 
rm -rf .
rm -rf test.*
rm -rf *.tex
cd ..
rm -rf 01
git status
cd src/
mv arithmetic/scala/ArithmeticExpression.scala main
mkdir arithmetic
rmdir arithmetic/ 
cd src/main/scala/
mkdir arithmetic
mv ArithmeticExpression.scala 

config.el
(setq gc-cons-threshold (* 50 1000 1000))

(defun efs/display-startup-time ()
  (message "Emacs loaded in %s with %d garbage collections."
           (format "%.2f seconds"
                   (float-time
                     (time-subtract after-init-time before-init-time)))
           gcs-done))

(add-hook 'emacs-startup-hook #'efs/display-startup-time)

(require 'package)

(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                         ("org" . "https://orgmode.org/elpa/")
                         ("elpa" . "https://elpa.gnu.org/packages/")))

(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))

  ;; Initialize use-package on non-Linux platforms
(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)

;; Disable backup files.
(setq make-backup-files nil)
;; Prompt to delete autosaves when killing buffers.
(setq kill-buffer-delete-auto-save-files t)

(global-visual-line-mode 0)
(toggle-truncate-lines 0)
(setq inhibit-splash-screen t)
(setq warning-minimum-level :error)  ; Only show errors, not warnings
(setq warning-suppress-types '((comp)))  ; Suppress all compilation warnings

;; magit
(keymap-global-set "C-c g" 'magit)

;; eshell
(keymap-global-set "C-c e" 'eshell)

;; buffer
;(which-key-add-key-based-replacements "C-c b" "buffer")

(keymap-global-set "C-c b i" 'ibuffer)
(keymap-global-set "C-c b p" 'previous-buffer)
(keymap-global-set "C-c b n" 'next-buffer)

;; Text scale
;(which-key-add-key-based-replacements "C-c s" "scale")
(keymap-global-set "C-c s i" 'text-scale-increase)
(keymap-global-set "C-c s d" 'text-scale-decrease)

;; compile
(keymap-global-set "C-c c" 'compile)

;; tooling
;(which-key-add-key-based-replacements "C-c t" "tools")
(keymap-global-set "C-c t s" 'sbt-start)
(keymap-global-set "C-c t n" 'sbt-switch-to-active-sbt-buffer)

(keymap-global-set "C-c r" 'rename-file)



(use-package diminish)
(diminish 'projectile-mode)

(use-package all-the-icons
  :ensure t
  :if (display-graphic-p))

(use-package all-the-icons-dired
  :hook (dired-mode . (lambda () (all-the-icons-dired-mode t))))
(add-hook 'dired-mode-hook 'dired-omit-mode)

(use-package auctex
  :defer t
  :mode ("\\.tex\\'" . tex-mode)
  :hook (tex-mode . lsp-deferred)
  :config
  (setq TeX-source-correlate-start-server t))

;; Make sure this path matches the one installed via cargo
(setenv "PATH" (concat (expand-file-name "~/.cargo/bin") ":" (getenv "PATH")))
(add-to-list 'exec-path (expand-file-name "~/.cargo/bin"))

(use-package smartparens
  :init
  (smartparens-global-mode))

(use-package move-text
:ensure t
:config
(move-text-default-bindings))  ;; binds M-↑ and M-↓

(use-package dired-open
  :config
  (setq dired-open-extensions '(("gif" . "sxiv")
                                ("jpg" . "sxiv")
                                ("png" . "sxiv")
                                ("mkv" . "mpv")
                                ("mp4" . "mpv"))))

(use-package flycheck
  :ensure t
  :init (global-flycheck-mode)
  :config
  (setq flycheck-display-errors-function
    #'flycheck-display-error-messages-unless-error-list)

  (setq flycheck-indication-mode nil))

(use-package flycheck-pos-tip
  :ensure t
  :after flycheck
  :config
  (flycheck-pos-tip-mode))

(defun make-shell (name)
  "Create a shell buffer named NAME."
  (interactive "sName: ")
  (setq name (concat "$" name))
  (eshell)
  (rename-buffer name))

(use-package nerd-icons)

(set-frame-font "FiraCode Nerd Font 19")
(set-face-attribute 'default nil
                    :font "FiraCode Nerd Font"
                    :height 110
                    :weight 'medium)
(set-face-attribute 'variable-pitch nil
                    :font "FiraCode Nerd Font"
                    :height 120
                    :weight 'medium)
(set-face-attribute 'fixed-pitch nil
                    :font "FiraCode Nerd Font"
                    :height 110
                    :weight 'medium)
;; Makes commented text and keywords italics.
;; This is working in emacsclient but not emacs.
;; Your font must have an italic face available.
(set-face-attribute 'font-lock-comment-face nil
                    :slant 'italic)
(set-face-attribute 'font-lock-keyword-face nil
                    :slant 'italic)

;; This sets the default font on all graphical frames created after restarting Emacs.
;; Does the same thing as 'set-face-attribute default' above, but emacsclient fonts
;; are not right unless I also add this method of setting the default font.
(add-to-list 'default-frame-alist '(font . "FiraCode Nerd Font 12"))

;; Uncomment the following line if line spacing needs adjusting.
(setq-default line-spacing 0.12)

(use-package magit
  :diminish)

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)

(use-package gradle-mode)

(use-package highlight-indent-guides
:config
(set-face-background 'highlight-indent-guides-odd-face "darkgray")
(set-face-background 'highlight-indent-guides-even-face "dimgray")
(set-face-foreground 'highlight-indent-guides-character-face "dimgray")
(add-hook 'c++-mode-hook 'highlight-indent-guides-mode)
(add-hook 'java-mode-hook 'highlight-indent-guides-mode)
(add-hook 'scala-mode-hook 'highlight-indent-guides-mode)
(add-hook 'prog-mode-hook 'highlight-indent-guides-mode))

(use-package hl-todo
  :hook ((org-mode . hl-todo-mode)
         (prog-mode . hl-todo-mode))
  :config
  (setq hl-todo-highlight-punctuation ":"
        hl-todo-keyword-faces
        `(("TODO"       warning bold)
          ("FIXME"      error bold)
          ("HACK"       font-lock-constant-face bold)
          ("REVIEW"     font-lock-keyword-face bold)
          ("NOTE"       success bold)
          ("DEPRECATED" font-lock-doc-face bold))))

(setq-default tab-width 4)
(setq-default indent-tabs-mode nil)

(use-package counsel
  :bind (("C-x b" . 'counsel-ibuffer)
         :map minibuffer-local-map
         ("C-r" . 'counsel-minibuffer-history))
  :custom
  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
  :config
  (counsel-mode 1))

(use-package lsp-mode
    :ensure t
    :commands (lsp lsp-deferred)
    :init
    (setq lsp-keymap-prefix "C-c l"
          lsp-modeline-diagnostics-enable nil)
    :hook ((LaTeX-mode . lsp-deferred)
           (lsp-mode . lsp-enable-which-key-integration)
           (julia-mode . lsp)
           (c-mode . lsp)
           (c++-mode . lsp)
           (C++-mode . lsp)
           (java-mode . lsp)
           (sh-mode . lsp)
           (css-mode . lsp)
           (tex-mode . lsp))
    :custom
    (lsp-rust-analyzer-cargo-watch-command "clippy")
    (lsp-eldoc-render-all t)
    (lsp-idle-delay 0.6)
    (lsp-inlay-hint-enable t)
    (lsp-log-io t)
    (lsp-diagnostics-provider :flycheck) ;; Explicitly use Flycheck
    :config
    (setq lsp-rust-analyzer-display-lifetime-elision-hints-enable t
          lsp-rust-analyzer-display-chaining-hints t
          lsp-rust-analyzer-display-lifetime-elision-hints-use-parameter-names nil
          lsp-rust-analyzer-display-closure-return-type-hints t
          lsp-rust-analyzer-display-parameter-hints nil
          lsp-rust-analyzer-display-reborrow-hints nil))

  (use-package lsp-latex
    :ensure t
    :hook (bibtex-mode . lsp))

  (use-package lsp-ui
    :ensure t
    :hook (lsp-mode . lsp-ui-mode)
    :custom
    (lsp-ui-doc-position 'bottom))

  (use-package lsp-treemacs
    :ensure t
    :after lsp)

  (use-package lsp-ivy
    :ensure t
    :after lsp)

(with-eval-after-load 'lsp-mode
  (define-key lsp-mode-map (kbd "TAB") nil))

(use-package company
    :after lsp-mode
    :hook (lsp-mode . company-mode)
    :bind (:map company-active-map
           ("C-u" . company-complete-selection)
           ("C-j" . company-select-next)
           ("C-k" . company-select-previous))
          (:map lsp-mode-map
           ("C-i" . company-indent-or-complete-common)
           ("<TAB>" . indent-for-tab-command))
    :custom
    (company-minimum-prefix-length 1)
    (company-idle-delay 0.0))

  (use-package company-box
    :ensure t
    :after company
    :hook (company-mode . company-box-mode))

(with-eval-after-load 'company
  (define-key company-active-map (kbd "TAB") nil)
  (define-key company-active-map (kbd "<tab>") nil))

(use-package lsp-java
  :hook (java-mode . lsp-deferred))

(use-package python-mode
    :ensure t
    :hook (python-mode . lsp-deferred)
    :custom
    ;; NOTE: Set these if Python 3 is called "python3" on your system!
    ;; (python-shell-interpreter "python3")
    ;; (dap-python-executable "python3")
    (dap-python-debugger 'debugpy)
    :config
    (require 'dap-python))

;; Pyright lsp configuration 
(use-package lsp-pyright
:ensure t
:custom (lsp-pyright-langserver-command "pyright") ;; or basedpyright
:hook (python-mode . (lambda ()
                        (require 'lsp-pyright)
                        (lsp))))  ; or lsp-deferred
;; Pytest
(use-package pytest)
(use-package pyvenv)
(use-package poetry
  :hook (python-mode . poetry-tracking-mode))

(use-package lsp-julia
       :after lsp-mode
        :config
(setq lsp-julia-default-environment "~/.julia/environments/v1.11"))

(use-package verilog-mode
  :ensure t
  :hook (verilog-mode . (lambda ()
                          (require 'verilog-mode)
                          (lsp))))

(use-package ccls
:ensure t
:hook ((c-mode c++-mode objc-mode cuda-mode) .
       (lambda () (require 'ccls) (lsp)))
:config
(progn
  (setq ccls-executable "/usr/bin/ccls")
  (setq ccls-initialization-options
        '(:index (:comments 2) :completion (:detailedLabel t)))))

(use-package haskell-mode
  :hook (haskell-mode . (lambda () (flycheck-mode -1))))

(use-package scala-mode
  :mode "\\.s\\(cala\\|bt\\)$"
  :hook (scala-mode . (lambda () (flycheck-mode -1) (electric-indent-mode -1))))

(use-package sbt-mode
  :commands sbt-start sbt-command
  :config
  ;; WORKAROUND: allows using SPACE when in the minibuffer
  (substitute-key-definition
   'minibuffer-complete-word
   'self-insert-command
   minibuffer-local-completion-map))

(use-package lsp-metals
  :ensure t
  :custom
  ;; You might set metals server options via -J arguments. This might not always work, for instance when
  ;; metals is installed using nix. In this case you can use JAVA_TOOL_OPTIONS environment variable.
  (lsp-metals-server-args '(;; Metals claims to support range formatting by default but it supports range
                            ;; formatting of multiline strings only. You might want to disable it so that
                            ;; emacs can use indentation provided by scala-mode.
                            "-J-Dmetals.allow-multiline-string-formatting=off"
                            ;; Enable unicode icons. But be warned that emacs might not render unicode
                            ;; correctly in all cases.
                            "-J-Dmetals.icons=unicode"))
  ;; In case you want semantic highlighting. This also has to be enabled in lsp-mode using
  ;; `lsp-semantic-tokens-enable' variable. Also you might want to disable highlighting of modifiers
  ;; setting `lsp-semantic-tokens-apply-modifiers' to `nil' because metals sends `abstract' modifier
  ;; which is mapped to `keyword' face.
  (lsp-metals-enable-semantic-highlighting t)
  :hook (scala-mode . lsp))

(setq package-selected-packages 
  '(dart-mode lsp-mode lsp-dart lsp-treemacs flycheck company
    ;; Optional packages
    lsp-ui company hover))

(when (cl-find-if-not #'package-installed-p package-selected-packages)
  (package-refresh-contents)
  (mapc #'package-install package-selected-packages))

(add-hook 'dart-mode-hook 'lsp)

(setq gc-cons-threshold (* 100 1024 1024)
      read-process-output-max (* 1024 1024))

(global-set-key [escape] 'keyboard-escape-quit)

(use-package toc-org
  :diminish
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable)
  (setq org-agenda-start-on-weekday 1)
  (setq org-agenda-files (list "~/University/uni.org" "~/thesis/notes.org")))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

(require 'org-tempo)

(use-package multiple-cursors
  :ensure t
  :bind (("C-S-c C-S-c" . mc/edit-lines)
         ("C->"         . mc/mark-next-like-this)
         ("C-<"         . mc/mark-previous-like-this)
         ("C-c C-<"     . mc/mark-all-like-this)))

(use-package pdf-tools
  :hook (pdf-view-mode . (lambda () (display-line-numbers-mode -1)))
  :config
  (pdf-tools-install))

(setq TeX-view-program-selection
    '((output-pdf "PDF Tools")))

(use-package org-noter
  :after pdf-tools
  :config
  (require 'org-noter-pdftools))

(use-package org-pdftools
  :hook (org-mode . org-pdftools-setup-link))

(use-package org-noter-pdftools
  :after org-noter
  :config
  ;; Add a function to ensure precise note is inserted
  (defun org-noter-pdftools-insert-precise-note (&optional toggle-no-questions)
    (interactive "P")
    (org-noter--with-valid-session
     (let ((org-noter-insert-note-no-questions (if toggle-no-questions
                                                   (not org-noter-insert-note-no-questions)
                                                 org-noter-insert-note-no-questions))
           (org-pdftools-use-isearch-link t)
           (org-pdftools-use-freepointer-annot t))
       (org-noter-insert-note (org-noter--get-precise-info)))))

  ;; fix https://github.com/weirdNox/org-noter/pull/93/commits/f8349ae7575e599f375de1be6be2d0d5de4e6cbf
  (defun org-noter-set-start-location (&optional arg)
    "When opening a session with this document, go to the current location.
With a prefix ARG, remove start location."
    (interactive "P")
    (org-noter--with-valid-session
     (let ((inhibit-read-only t)
           (ast (org-noter--parse-root))
           (location (org-noter--doc-approx-location (when (called-interactively-p 'any) 'interactive))))
       (with-current-buffer (org-noter--session-notes-buffer session)
         (org-with-wide-buffer
          (goto-char (org-element-property :begin ast))
          (if arg
              (org-entry-delete nil org-noter-property-note-location)
            (org-entry-put nil org-noter-property-note-location
                           (org-noter--pretty-print-location location))))))))
  (with-eval-after-load 'pdf-annot
    (add-hook 'pdf-annot-activate-handler-functions #'org-noter-pdftools-jump-to-note)))

(use-package perspective
  :custom
  ;; NOTE! I have also set 'SCP =' to open the perspective menu.
  ;; I'm only setting the additional binding because setting it
  ;; helps suppress an annoying warning message.
  (persp-mode-prefix-key (kbd "C-c M-p"))
  :init 
  (persp-mode)
  :config
  ;; Sets a file to write to when we save states
  (setq persp-state-default-file "~/.config/emacs/sessions"))

;; This will group buffers by persp-name in ibuffer.
(add-hook 'ibuffer-hook
          (lambda ()
            (persp-ibuffer-set-filter-groups)
            (unless (eq ibuffer-sorting-mode 'alphabetic)
              (ibuffer-do-sort-by-alphabetic))))

;; Automatically save perspective states to file when Emacs exits.
(add-hook 'kill-emacs-hook #'persp-state-save)

(use-package rainbow-delimiters
  :hook ((emacs-lisp-mode . rainbow-delimiters-mode)
         (clojure-mode . rainbow-delimiters-mode)))

(use-package rainbow-mode
  :diminish
  :hook 
  ((org-mode prog-mode) . rainbow-mode))

(use-package eshell-syntax-highlighting
  :after esh-mode
  :config
  (eshell-syntax-highlighting-global-mode +1))

;; eshell-syntax-highlighting -- adds fish/zsh-like syntax highlighting.
;; eshell-rc-script -- your profile for eshell; like a bashrc for eshell.
;; eshell-aliases-file -- sets an aliases file for the eshell.

(setq eshell-rc-script (concat user-emacs-directory "eshell/profile")
      eshell-aliases-file (concat user-emacs-directory "eshell/aliases")
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "fish" "htop" "ssh" "top" "zsh"))

(use-package doom-themes
  :ensure t
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
        doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-snazzy t)

  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)
  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-atom") ; use "doom-colors" for less minimal icon theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))

(use-package tree-sitter
      :diminish
      :init
  (global-tree-sitter-mode)
  (add-hook 'tree-sitter-after-on-hook #'tree-sitter-hl-mode))
(use-package tree-sitter-langs
    :diminish)

(use-package which-key
  :init
  (which-key-mode 1)
  :config
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 6
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.8
        which-key-max-description-length 25
        which-key-allow-imprecise-window-fit nil 
        which-key-separator " → " ))

(use-package yasnippet
  :ensure t
  :hook ((LaTeX-mode . yas-minor-mode)
         (post-self-insert . my/yas-try-expanding-auto-snippets)))

(setq yas-triggers-in-field t)

;; Function that tries to autoexpand YaSnippets
;; The double quoting is NOT a typo!
(defun my/yas-try-expanding-auto-snippets ()
  (when (and (boundp 'yas-minor-mode) yas-minor-mode)
    (let ((yas-buffer-local-condition ''(require-snippet-condition . auto)))
      (yas-expand))))

(defun unbind-shift-keybindings ()
  "Unbind all keybindings containing the Shift key."
  (interactive)
  (let ((shift-key-prefixes '("S-" "s-")))
    (mapc (lambda (keymap)
            (mapc (lambda (prefix)
                    (mapc (lambda (key)
                            (define-key keymap (kbd (concat prefix key)) nil))
                          '("!" "\"" "#" "$" "%" "&" "'" "(" ")" "*"
                            "+" "," "-" "." "/" ":" ";" "<" "=" ">" "?"
                            "@" "[" "\\" "]" "^" "_" "`" "{" "|" "}" "~"
                            "0" "1" "2" "3" "4" "5" "6" "7" "8" "9"
                            "A" "B" "C" "D" "E" "F" "G" "H" "I" "J"
                            "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T"
                            "U" "V" "W" "X" "Y" "Z")))
                  shift-key-prefixes))
          (list global-map (current-local-map)))))
